### PR TITLE
Clinical RFC

### PR DESCRIPTION
**Pull request by aditigopalan**
# Add Clinical Module to Data Model

## Changes Made
- Added Clinical module with comprehensive schema for clinical data
- Added test data demonstrating schema usage
- Enhanced Participant module with additional demographic fields
- Implemented comprehensive test suite for both Clinical and Participant modules
- Added detailed development documentation and quickstart guides
- Improved build system and CI workflow
- Fixed schema path dependencies in Makefile

## Schema Features
- Defines clinical data classes including:
  - Diagnosis information
  - Exposure history
  - Family history
  - Follow-up observations
  - Molecular test results
  - Therapy information
- Includes comprehensive enums for standardized values
- Links to standard ontologies (NCIT, UBERON)
- Enhanced Participant schema with:
  - Demographic fields (race, ethnicity, gender, sex)
  - Vital status tracking
  - Physical measurements
  - Location information

## Documentation
- Added comprehensive development guide (`docs/development.md`) covering:
  - Local development setup
  - Testing procedures
  - CI workflow testing
  - Project structure
  - Common development tasks
  - Troubleshooting guide
  - Best practices
- Added quickstart guide (`docs/quickstart.md`) with:
  - Common commands
  - Setup instructions
  - Testing workflows
  - Common issues and solutions
  - Project structure overview

## Build System Improvements
- Updated Makefile to use correct schema paths
- Fixed example generation dependencies
- Improved module-specific build rules
- Enhanced test execution workflow
- Added proper documentation handling in .gitignore

## Testing
✅ Validated schema and test data using LinkML validator:
```bash
linkml-validate -s modules/Clinical/clinical_schema.yaml --target-class ClinicalData modules/Clinical/test_data.yaml
```
Test passed successfully with no validation issues, confirming:
- Schema syntax is correct
- Test data conforms to all schema constraints
- Import paths are working correctly

## Test Coverage
- Added unit tests for both Clinical and Participant modules
- Implemented validation tests for:
  - Required fields
  - Data type constraints
  - Enum value validation
  - Complex nested structures
- Added test data examples for:
  - Valid cases demonstrating proper usage
  - Invalid cases testing validation rules
  - Edge cases and error conditions
- All tests passing locally and in CI

## CI/CD Improvements
- Fixed CI workflow to properly handle schema paths
- Added local CI testing capabilities
- Improved test execution in CI environment
- Enhanced documentation generation workflow

## Future Considerations
- Consider adding more specific test cases for edge conditions
- May want to add documentation for common validation patterns
- Could expand test coverage for specific clinical scenarios
- Consider adding automated documentation generation
- May want to add more detailed schema validation rules

### CODE REVIEW

File: .github/workflows/deploy-docs.yaml [hunk 1]
```diff
@@ -31,10 +36,131 @@ jobs:
...
+          git config --local user.email "action@github.com"
+          git config --local user.name "GitHub Action"
+          git add docs/
+          git commit -m "docs: update generated documentation" || true
+          git push || true
+
+      - name: Build documentation.
+        run: |
+          # Build the documentation using Poetry
+          poetry run mkdocs build
```
Comment (thomasyu888) [inline .github/workflows/deploy-docs.yaml:157]: Consider using readthedocs to automatically deploy your documentation so you don't have to use gh-pages or maintain this CI/CD pipeline.  This is what we do in the python client. 

I recommend a "design" ticket to look into what this would be and explore readthedocs.

File: scripts/linkml_to_flat_synapse_jsonschema.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
+import subprocess
+import sys
+import json
+from pathlib import Path
+import jsonref
+
+def run_gen_json_schema(linkml_yaml, class_name, tmp_json):
```
Comment (thomasyu888) [inline scripts/linkml_to_flat_synapse_jsonschema.py:7]: Use type linking

```
def run_gen_json_schema(linkml_yaml: str, class_name: str, tmp_json: str)
```

Add docstrings and add Argument definitions.  So an example is this:


```
def get_assay_dataframe(filepath_list: list) -> pd.DataFrame:
    """Reads in assay_information.yaml file
        and outputs it as a dataframe

    Args:
        filepath_list (list): list of files

    Raises:
        ValueError: thrown if read error with file

    Returns:
        pd.DataFrame: dataframe version of assay info file
    """
```

File: scripts/linkml_to_flat_synapse_jsonschema.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    # Dereference $ref
+    deref_schema = jsonref.JsonRef.replace_refs(schema)
+    # Convert to plain dict to ensure JSON serializable
+    def convert_to_plain_dict(obj):
+        if isinstance(obj, dict):
+            return {k: convert_to_plain_dict(v) for k, v in obj.items()}
+        elif isinstance(obj, list):
+            return [convert_to_plain_dict(item) for item in obj]
+        else:
+            return obj
```
Comment (thomasyu888) [inline scripts/linkml_to_flat_synapse_jsonschema.py:28]: I'm personally not a huge fan of nested functions because you can't test them.  I would pull this out and make it a "hidden" and/or helper function with a `_` in front of it.

File: scripts/linkml_to_flat_synapse_jsonschema.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+            print(f"Updated $schema from '{current_schema}' to 'https://json-schema.org/draft-07/schema'")
+        else:
+            print(f"$schema already uses Draft-07: {current_schema}")
+    else:
+        # Add $schema if it doesn't exist
+        data["$schema"] = "https://json-schema.org/draft-07/schema"
+        print("Added $schema field with Draft-07")
+    with open(filepath, "w") as f:
+        json.dump(data, f, indent=2)
+    print(f"Fixed schema version in {filepath}")
```
Comment (thomasyu888) [inline scripts/linkml_to_flat_synapse_jsonschema.py:53]: Let's just take this function and talk about how you would test this.  It would require you to have an actual jsonschema file, BUT really what you want to test is this portion

```
    if "$schema" in data:
        current_schema = data["$schema"]
        if "draft-07" not in current_schema:
            data["$schema"] = "https://json-schema.org/draft-07/schema"
            print(f"Updated $schema from '{current_schema}' to 'https://json-schema.org/draft-07/schema'")
        else:
            print(f"$schema already uses Draft-07: {current_schema}")
    else:
        # Add $schema if it doesn't exist
        data["$schema"] = "https://json-schema.org/draft-07/schema"
        print("Added $schema field with Draft-07")
```

Therefore, this internal function could be pulled out into a separate helper function which would make it easier for you to test.

File: scripts/linkml_to_flat_synapse_jsonschema.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+def remove_unsupported_fields(filepath):
+    with open(filepath, "r") as f:
+        data = json.load(f)
+    def recursive_clean(obj):
+        if isinstance(obj, dict):
+            for field in ["$defs", "metamodel_version", "version"]:
+                if field in obj:
+                    del obj[field]
+            for v in obj.values():
```
Comment (thomasyu888) [inline scripts/linkml_to_flat_synapse_jsonschema.py:63]: `v` is not a descriptive parameter, consider renaming this to be more descriptive.

File: scripts/linkml_to_flat_synapse_jsonschema.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+                recursive_fix(item)
+    recursive_fix(data)
+    with open(filepath, "w") as f:
+        json.dump(data, f, indent=2)
+    print(f"Fixed additionalProperties in {filepath}")
+
+def main():
+    if len(sys.argv) < 2 or len(sys.argv) > 4:
+        print("Usage: python linkml_to_flat_synapse_jsonschema.py <linkml_yaml> [class_name] [output_filename]")
+        sys.exit(1)
```
Comment (thomasyu888) [inline scripts/linkml_to_flat_synapse_jsonschema.py:94]: Use argparse like you do for another script.

File: scripts/linkml_to_flat_synapse_jsonschema.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    # 4. Fix additionalProperties
+    fix_additional_properties(output_file)
+    # 5. Remove unsupported fields
+    remove_unsupported_fields(output_file)
+    # 6. Cleanup
+    Path(tmp_json).unlink(missing_ok=True)
+    print(f"✅ Synapse-compatible flat schema written to {output_file}")
+
+if __name__ == "__main__":
+    main()
```
Comment (thomasyu888) [inline scripts/linkml_to_flat_synapse_jsonschema.py:121]: Lastly, I want to say that I really like how you've laid out the workflow in the `main` function, it's very clear what happens between each of the steps.  If I had to give feedback is that I would read in the jsonschema once, and then write it out at the end. 

Right now you're reading and writing the jsonschema at each step, but that may not be necessary.  One minimal risk is that something can also write to the same jsonschema that is being edited by this code.  Having the jsonschema in memory could mitigate that.

If there are memory overload issues due to the recursive functions, then consider using `tempfile` module to create temporary directory or temporary files.  This way you don't need to do this: `Path(tmp_json).unlink(missing_ok=True)`

File: minimal_test_schema.yaml
Comment (adamjtaylor) [inline minimal_test_schema.yaml:1]: Hs this fole been removed or just made empty? Should it be in the root dir?

**Review by adamjtaylor — COMMENTED** [review]
There are **a lot** of changes here. Its a huge ammount of work that really gets the flywheel going. 

I think we need to check in and make sure we've made this scalable for the future and not overcomplicated.

We should make sure it is clear which files are expected to be autogenerated and move them into post/within PR actions such that we only have to review changes to the root files and not the subsequently auto-generated ones.

Lets chat more!

File: convert_uberon.py
[the remark this answers is not in the document]
Comment (thomasyu888) [reply convert_uberon.py:1]: This could also be in a general utilities script if there is one so you aren't maintaining separate scripts. Consider having a small argparse interface or use something like typer or click to create simple utility CLI code.

File: docs/index.md
[the remark this answers is not in the document]
Comment (thomasyu888) [reply docs/index.md:1]: Agreed, i think we can review the docs separately once it's published and that review and it's comments can be done in a separate PR.

A quick meeting once the page is published would be easiest to walk through the docs and create tickets around this work.

File: mkdocs.yml
Comment (thomasyu888) [inline mkdocs.yml:1]: As a separate ticket, explore using readthedocs to auto generate the documentation. readthedocs also auto hosts it so you can remove that from the CI/CD pipeline all together

File: modules/Clinical/tests/test_conditional_requirements.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
+import pytest
+from pathlib import Path
+import sys
+import os
+import yaml
+from copy import deepcopy
+from linkml_runtime.loaders import yaml_loader
+from linkml_runtime.dumpers import yaml_dumper
+from htan_clinical.datamodel.clinical import ClinicalData
```
Comment (thomasyu888) [inline modules/Clinical/tests/test_conditional_requirements.py:9]: Use the isort package to auto sort these package dependencies. (pep8 standards for package import ordering)

File: modules/Clinical/tests/test_conditional_requirements.yaml
Comment (thomasyu888) [inline modules/Clinical/tests/test_conditional_requirements.yaml:1]: Is this test data?  consider moving these into a test_data directory within the tests folder for organization purposes but that's not totally needed.

I'll leave it to your discretion, but I would start making the distinction between smaller changes vs larger ones and create PR's into this large feature branch.

It's also ok to create tickets for these and merge this in so we can start with a clean slate.    It is a lot on reviewers to review such a large PR especially if lots of files are changing.

File: modules/Clinical/tests/test_clinical.py
Comment (thomasyu888) [inline modules/Clinical/tests/test_clinical.py:1]: Great usage of pytest!

File: pytest.ini
Comment (thomasyu888) [inline pytest.ini:1]: Could this just be included in the `pyproject.toml`? (This could be a ticket that you create)

Think agile - what is the acceptance criteria of this entire PR (there's probably many), but as indicated above, we want to get this to a state of "good enough" to ship, then let's start branching off and iterating in smaller chunks.

File: scripts/linkml_to_flat_synapse_jsonschema.py [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+    schema_str = generator.serialize()
+
+    # Write to temporary file
+    with open(tmp_json, "w") as f:
+        f.write(schema_str)
+
+    print(f"Generated JSON Schema written to {tmp_json}")
+
+
+def _convert_to_plain_dict(obj) -> Union[dict, list, Any]:
```
Comment (thomasyu888) [inline scripts/linkml_to_flat_synapse_jsonschema.py:58]: Do you have an example here?  This is kind of interesting, because you're stating that the `obj` that is passed in is either `dict` and/or `list` or something else which I would expect it to actually be a `jsonred.JsonRef` object?

File: scripts/push_annotations_from_files.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/env python3
+"""Push annotations from JSON files to Synapse entities"""
+
+import synapseclient
+import json
+
+def push_annotations_from_file(file_path, entity_id):
+    """Push annotations from a JSON file to a Synapse entity"""
+    
+    print(f"Pushing annotations from {file_path} to entity {entity_id}")
```
Comment (thomasyu888) [inline scripts/push_annotations_from_files.py:10]: A whole other ticket you can create, but experiment with the logger function in Python. https://www.dash0.com/guides/logging-in-python

File: scripts/push_annotations_from_files.py [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+        print(f"✅ Retrieved entity: {entity.name}")
+        
+        # Apply the annotations
+        # Convert single values to lists (Synapse expects lists)
+        synapse_annotations = {}
+        for key, value in annotations.items():
+            if isinstance(value, list):
+                synapse_annotations[key] = value
+            else:
+                synapse_annotations[key] = [value]
```
Comment (thomasyu888) [inline scripts/push_annotations_from_files.py:38]: Depending on what version of the client you are using, you may want to consider using the new OOP interface that we are moving towards as much of this stuff will be deprecated
